### PR TITLE
Make MVC plugin support inherited annotations from interface

### PR DESCRIPTION
**Pull request by dfkjadkfj**
- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [x] Improve performance

- Related issues
  #3847 how to enhance method with @RequestMapping annotation declared in interface?
___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import net.bytebuddy.matcher.CollectionItemMatcher;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import java.util.Objects;
+
+import static net.bytebuddy.matcher.ElementMatchers.annotationType;
+
+/**
+ * @description: Matching used to match method annotations, Can match annotations on interface methods
+ * @auther: jialong  by 2020-01-23 15:41
```
Comment (kezhenxu94) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java]: typo `@author`, and please remove the `:`

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import net.bytebuddy.description.type.TypeList;
+import net.bytebuddy.matcher.CollectionItemMatcher;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import java.util.Objects;
+
+import static net.bytebuddy.matcher.ElementMatchers.annotationType;
+
+/**
+ * @description: Matching used to match method annotations, Can match annotations on interface methods
```
Comment (kezhenxu94) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java]: remove `@description`, it's not a valid JavaDoc tag

File: apm-sniffer/apm-agent-core/src/test/java/org/apache/skywalking/apm/agent/core/plugin/bytebuddy/MethodInheritsDeclaringAnnotationMatcherTest.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.apache.skywalking.apm.agent.core.plugin.match.MethodInheritsDeclaringAnnotationMatcher.isInheritsAnnotatedWith;
+
+/**
+ * @description:
+ * @auther: jialong  by 2020-01-23 16:09
```
Comment (kezhenxu94) [inline apm-sniffer/apm-agent-core/src/test/java/org/apache/skywalking/apm/agent/core/plugin/bytebuddy/MethodInheritsDeclaringAnnotationMatcherTest.java]: ditto

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java (hunk 1 above)
Comment (dfkjadkfj) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java]: Thanks for your correction

**Comment by dfkjadkfj** [comment]
For this PR I updated the test case of spring-4.3.x-scenario, please review

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        for (int i = 0; i < source.size(); i++) {
+            if (!Objects.equals(source.get(i).getType(), impl.get(i).getType())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static <T extends AnnotationSource> ElementMatcher.Junction<T> isInheritsAnnotatedWith(ElementMatcher<? super TypeDescription> matcher) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java]: Inherit is a verb. This method name should be `byMethodInheritanceAnnotationMatcher`

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Objects;
+
+import static net.bytebuddy.matcher.ElementMatchers.annotationType;
+
+/**
+ * Matching used to match method annotations, Can match annotations on interface methods
+ * @auther jialong
+ */
+@HashCodeAndEqualsPlugin.Enhance
+public class MethodInheritsDeclaringAnnotationMatcher<T extends MethodDescription> extends ElementMatcher.Junction.AbstractBase<T> {
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritsDeclaringAnnotationMatcher.java]: Should rename to `MethodInheritanceAnnotationMatcher`

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritanceAnnotationMatcher.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import net.bytebuddy.matcher.CollectionItemMatcher;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import java.util.Objects;
+
+import static net.bytebuddy.matcher.ElementMatchers.annotationType;
+
+/**
+ * Matching used to match method annotations, Can match annotations on interface methods
+ * @auther jialong
```
Comment (kezhenxu94) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritanceAnnotationMatcher.java]: Still, `@auther` is not a valid JavaDoc

```suggestion
 * @author jialong
```

**Comment by wu-sheng** [comment]
```
[INFO] Running org.apache.skywalking.apm.agent.core.plugin.bytebuddy.MethodInheritanceAnnotationMatcherTest
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.09 s <<< FAILURE! - in org.apache.skywalking.apm.agent.core.plugin.bytebuddy.MethodInheritanceAnnotationMatcherTest
[ERROR] testMatch(org.apache.skywalking.apm.agent.core.plugin.bytebuddy.MethodInheritanceAnnotationMatcherTest)  Time elapsed: 0.09 s  <<< FAILURE!
java.lang.AssertionError
	at org.apache.skywalking.apm.agent.core.plugin.bytebuddy.MethodInheritanceAnnotationMatcherTest.testMatch(MethodInheritanceAnnotationMatcherTest.java:44)
```

Test fail, please recheck. https://github.com/apache/skywalking/pull/4278/checks?check_run_id=420562649

**Comment by dfkjadkfj** [comment]
Sorry for not communicating in a timely manner, I have found a mistake, and am now fixing it, and returned to the countryside in the New Year, the network is not very good

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritanceAnnotationMatcher.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+
+    private boolean recursiveMatches(TypeDefinition typeDefinition, String methodName, ParameterList<?> parameters) {
+        TypeList.Generic interfaces = typeDefinition.getInterfaces();
+        for (TypeDescription.Generic implInterface : interfaces) {
+            if (recursiveMatches(implInterface, methodName, parameters)) {
+                return true;
+            }
+            MethodList<MethodDescription.InGenericShape> declaredMethods = implInterface.getDeclaredMethods();
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/match/MethodInheritanceAnnotationMatcher.java]: Is this right? Should this be out of `for` loop?

**Comment by wu-sheng** [comment]
Welcome to be the 201st contributor in this repo.

**Pull request merged**